wai-extra/Network/Wai/Middleware/Approot.hs

```diff
@@ -0,0 +1,120 @@
+import           Network.Wai.Request   (appearsSecure)
+import           System.Environment    (getEnvironment)
+import           System.IO.Unsafe      (unsafePerformIO)
+
+approotKey :: V.Key ByteString
+approotKey = unsafePerformIO V.newKey
+{-# NOINLINE approotKey #-}
+
+-- | The most generic version of the middleware, allowing you to provide a
+-- function to get the approot for each request. For many use cases, one of the
+-- helper functions provided by this module will give the necessary
+-- functionality more conveniently.
+--
+-- Since 3.0.7
+approotMiddleware :: (Request -> IO ByteString) -- ^ get the approot
+                  -> Middleware
+approotMiddleware getRoot app req respond = do
+    ar <- getRoot req
+    let req' = req { vault = V.insert approotKey ar $ vault req }
+    app req' respond
+
+-- | Same as @'envFallbackNamed' "APPROOT"@.
+--
+-- The environment variable @APPROOT@ is used by Keter, School of Haskell, and yesod-devel.
+--
+-- Since 3.0.7
+envFallback :: IO Middleware
+envFallback = envFallbackNamed "APPROOT"
+
+-- | Produce a middleware that takes the approot from the given environment
+-- variable, falling back to the behavior of 'fromRequest' if the variable is
+-- not set.
+--
+-- Since 3.0.7
+envFallbackNamed :: String -> IO Middleware
+envFallbackNamed name = do
+    env <- getEnvironment
+    case lookup name env of
+        Just s -> return $ hardcoded $ S8.pack s
+        Nothing -> return fromRequest
+
+-- | Hard-code the given value as the approot.
+--
+-- Since 3.0.7
+hardcoded :: ByteString -> Middleware
+hardcoded ar = approotMiddleware (const $ return ar)
+
+-- | Get the approot by analyzing the request. This is not a full-proof
+-- approach, but in many common cases will work. Situations that can break this
+-- are:
+--
+-- * Requests which spoof headers and imply the connection is over HTTPS
+--
+-- * Reverse proxies that change ports in surprising ways
+--
+-- * Invalid Host headers
+--
+-- * Reverse proxies which modify the path info
+--
+-- Normally trusting headers in this way is insecure, however in the case of
+-- approot, the worst that can happen is that the client will get an incorrect
+-- URL. If you are relying on the approot for some security-sensitive purpose,
+-- it is highly recommended to use @hardcoded@, which cannot be spoofed.
+--
+-- Since 3.0.7
+fromRequest :: Middleware
+fromRequest = approotMiddleware (return . extract)
+
+extract :: Request -> ByteString
+extract req =
+    (if appearsSecure req then "https://" else "http://") `S.append`
+    (fromMaybe "localhost" (requestHeaderHost req))
+
+data ApprootMiddlewareNotSetup = ApprootMiddlewareNotSetup
+    deriving (Show, Typeable)
+instance Exception ApprootMiddlewareNotSetup
+
+-- | Get the approot set by the middleware. If the middleware is not in use,
+-- then this function will return an exception.
+--
+-- Since 3.0.7
+getApproot :: Request -> ByteString
```

Comment:
I preferred exceptions because in the use cases I was considering, a `Nothing` should never happen. A `Nothing` only occurs in the case of user error (not setting up the middleware). However, now that you raise it, I suppose an application could speculatively check for the presence of the middleware and, if not present, do something else. So now my thought is: let's provide two versions, `getApproot` as defined, and `getApprootMay`, with a short explanation in the comments along what I've just written.